module_info_import/models/module_information.py

```diff
@@ -1,107 +1,92 @@
-            "name": module_info["name"],
-            "description_rst": module_info["description_rst"],
-            "description_html": module_info["description_rst_html"],
-            "authors": module_info["author_ids_description"],
+    def _update_or_create_modules(self, version, orga_name, repo_name, module):
+        odoo_version_dct = {
+            v.version: v.id for v in self.env["odoo.version"].search([])
+        }
+        repo = self._get_or_create_repo(orga_name, repo_name)
+        version_id = odoo_version_dct.get(version, "odoo_version_unknown")
+        vals = {
+            "available_version_ids": [(4, version_id, 0)],
+            "repo_id": repo.id,
+            "technical_name": module,
         }
+        module = self.search(
+            [("technical_name", "=", module), ("partner_id", "=", False)]
+        )
+        if module:
+            if module._should_update_module(version, orga_name):
+                module.write(vals)
+            module._add_available_version(version)
+        else:
+            module = self.create(vals)
 
     @api.model
-    def _update_modules(self, data):
-        # Since the goal is to resynchronize all modules and versions, it
-        # seems appropriate to make one big search and then work with cache.
-        #        existing_module = self.search([])
-        #        test = dict([(x.name, x.id) for x in existing_module])
-        #        existing_module_names = test.keys()
-        #        for module_info in data:
-        #            if module_info['technical_name'] in existing_module_names:
-        #                module_id = test.get(module_info['technical_name'])
-        #                self.browse(module_id).write({'name': module_info['technical_name']})
-        #            else:
-        #                self.create({'name': {'name': module_info['technical_name']}})
-        odoo_version_dct = dict(
-            [(v.version, v.id) for v in self.env["odoo.version"].search([])]
+    def _get_or_create_repo(self, orga_name, repo_name):
+        repo_vals = {"organization": orga_name, "name": repo_name}
+        repo = self.env["module.repo"].search(
+            [("name", "=", repo_name), ("organization", "=", orga_name)], limit=1
         )
+        if repo and repo.organization != orga_name:
+            self.env["module.repo"].create(repo_vals)
+        elif not repo:
+            repo = self.env["module.repo"].create(repo_vals)
+        return repo
 
-        for module_info in data:
-            module = self.search(
-                [("technical_name", "=", module_info["technical_name"])]
+    @api.model
+    def _add_available_version(self, version):
+        if version not in self.mapped("available_version_ids.version"):
+            self.env["module.version"].create(
+                {
+                    "module_id": self.id,
+                    "version_id": self._get_version(),
+                    "state": "done",
+                }
```

Comment:
Be careful version can existing from an existing PR migration

- note version should be an odoo.version object

```suggestion
    def _add_available_version(self, version):
        module_version = self.module_version_ids.filtered(lambda s: s.version_id == version)
        if module_version and module_version.state != "done":
            # In case that a version have been creating from a pending PR
            module_version.state = "done"
        else:
            self.env["module.version"].create(
                {
                    "module_id": self.id,
                    "version_id": version.id,
                    "state": "done",
                }
```